### PR TITLE
Add tests for 2i inputs to map reduce

### PR DESCRIPTION
**Pull request by russelldb**

### CODE REVIEW

File: src/mapred_verify.erl [hunk 1]
```diff
@@ -153,13 +171,89 @@ verify_filter_job(Client, Bucket, KeyCount, JobDesc, Verifier) ->
...
+    case is_2i_backend(Node, Bucket) of
+        true ->
+            Range = {index, Bucket, <<"$bucket">>, Bucket},
+            Start = erlang:now(),
+            {_, Node, _} = Client,
+            {ok, Result} = rpc:call(Node, riak_kv_mrc_pipe, mapred, [Range, JobDesc]),
+            End = erlang:now(),
+            {mapred_verifiers:Verifier(bucket, Result, KeyCount),
+             erlang:round(timer:now_diff(End, Start) / 1000)};
+        false -> {true, 0}
```
Comment (beerriot) [inline src/mapred_verify.erl]: Should we log some sort of "skipping 2i test on non-2i backend" warning here, to keep us from accidentally believing that "0" actually passed a test?

Comment (russelldb) [inline src/mapred_verify.erl]: I guess, really I should generalise this pattern to a fun. Lemme refactor a bit.

**Comment by beerriot** [comment]
The failure I get without the patch in basho/riak_kv#618 appears to be in transcoding the `{Bucket, {Key, IndexTerm}}` to a Javascript-usable format. Is that what you see as well?

If so, +1 new tests look great.

Do we need to file an issue against basho/giddyup to make sure `mapred_verify_rt` is run against a 2i backend?

**Comment by russelldb** [comment]
I get this:

```
{{badmatch,
 {error,
     {1,
      [{module,riak_kv_w_reduce},
       {partition,959110449498405040071168171470060731649205731328},
       {details,
           [{fitting,
                {fitting,<10344.9048.0>,#Ref<10344.0.0.61742>,
                    <<"????????????????????">>,1}},
            {name,1},
            {module,riak_kv_w_reduce},
            {arg,{rct,#Fun<riak_kv_w_reduce.0.20542221>,none}},
            {output,
                {fitting,<10344.9046.0>,#Ref<10344.0.0.61742>,sink,
                    undefined}},
            {options,
                [{log,sink},
                 {trace,[error]},
                 {sink,
                     {fitting,<10344.9046.0>,#Ref<10344.0.0.61742>,sink,
                         undefined}},
                 {sink_type,{fsm,10,infinity}}]},
            {q_limit,64}]},
       {type,exit},
       {error,bad_json},
       {input,
           {not_found,{<<"mr_validate">>,{496,<<"mrv496">>}},undefined}},
       {modstate,
           {state,[],
               [{not_found,
                    {<<"mr_validate">>,{438,<<"mrv438">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{458,<<"mrv458">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{468,<<"mrv468">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{500,<<"mrv500">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{491,<<"mrv491">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{495,<<"mrv495">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{497,<<"mrv497">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{493,<<"mrv493">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{487,<<"mrv487">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{471,<<"mrv471">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{498,<<"mrv498">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{499,<<"mrv499">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{466,<<"mrv466">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{473,<<"mrv473">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{476,<<"mrv476">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{470,<<"mrv470">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{445,<<"mrv445">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{474,<<"mrv474">>}},
                    undefined},
                {not_found,
                    {<<"mr_validate">>,{489,<<"mrv489">>}},
                    undefined}],
               19,20,959110449498405040071168171470060731649205731328,
               {fitting_details,
                   {fitting,<10344.9048.0>,#Ref<10344.0.0.61742>,
                       <<"????????????????????">>,1},
                   1,riak_kv_w_reduce,
                   {rct,#Fun<riak_kv_w_reduce.0.20542221>,none},
                   {fitting,<10344.9046.0>,#Ref<10344.0.0.61742>,sink,
                       undefined},
                   [{log,sink},
                    {trace,[error]},
                    {sink,
                        {fitting,<10344.9046.0>,#Ref<10344.0.0.61742>,
                            sink,undefined}},
                    {sink_type,{fsm,10,infinity}}],
                   64}}},
       {stack,
           [{riak_kv_w_reduce,'-js_runner/1-fun-0-',3,
                [{file,"src/riak_kv_w_reduce.erl"},{line,283}]},
            {riak_kv_w_reduce,reduce,3,
                [{file,"src/riak_kv_w_reduce.erl"},{line,206}]},
            {riak_kv_w_reduce,maybe_reduce,2,
                [{file,"src/riak_kv_w_reduce.erl"},{line,157}]},
            {riak_pipe_vnode_worker,process_input,3,
                [{file,"src/riak_pipe_vnode_worker.erl"},{line,445}]},
            {riak_pipe_vnode_worker,wait_for_input,2,
                [{file,"src/riak_pipe_vnode_worker.erl"},{line,377}]},
            {gen_fsm,handle_msg,7,[{file,"gen_fsm.erl"},{line,494}]},
            {proc_lib,init_p_do_apply,3,
                [{file,"proc_lib.erl"},{line,227}]}]}]}}},
 [{mapred_verify,verify_2i_range_job,5,
  [{file,"src/mapred_verify.erl"},{line,251}]},
  {mapred_verify,'-verify_job/6-fun-13-',7,
  [{file,"src/mapred_verify.erl"},{line,130}]},
  {lists,foldl,3,[{file,"lists.erl"},{line,1197}]},
  {mapred_verify,'-run_jobs/4-fun-0-',5,
  [{file,"src/mapred_verify.erl"},{line,93}]},
  {lists,foldl,3,[{file,"lists.erl"},{line,1197}]},
  {mapred_verify_rt,confirm,0,
  [{file,"tests/mapred_verify_rt.erl"},{line,53}]}]}
```

Yes an JSON error, but also, not founds, since the `key` is not a `key` but a pair `{index_term, key}`

This matches the error report here http://lists.basho.com/pipermail/riak-users_lists.basho.com/2013-July/012777.html (which I can reproduce with the riak-erlang-client.)

**Comment by beerriot** [comment]
Sweet. That looks exactly like what I got as well.

**Comment by russelldb** [comment]
I added another commit. Spotted that one of the older tests was checking for 2i status, oops.
Also added logging of skipped tests.

**Comment by beerriot** [comment]
:P embarrassed that I also missed that. Good catch.

I've re-verified that the tests work when they're supposed to, break when their supposed to, and skip (and now warn) when they're supposed to. Re +1.

**Comment by russelldb** [comment]
Killing as part of an "my old" PRs cleanup

**Pull request closed without merging**